lib/prism/parse_result.rb

```diff
@@ -219,7 +215,8 @@ def join(other)
     # the beginning of the file. Useful for when you want a location object but
     # do not care where it points.
     def self.null
-      new(nil, 0, 0)
+      source = nil #: Source
+      new(source, 0, 0)
```

Comment:
So this one's tricky. I chose to type `Location#source` as `Source`, not `Source | nil`. That let me avoid a lot of runtime nil checks, but on the flip side required unsound assertions here in `Location.null` and also in `DSL#Location`. Both places felt like edge-case usages, so I went with that.

Do you have a preference?

Reply:
IMO the type should represent the reality, so Location#source can be nil and in fact it's not hard to see objects which have that when building new nodes with the DSL, which uses Location.null as default.

Reply:
Addressed in ac0af0533b59e8ffe21f4a4cfb06c4447ec82174